File: app/src/main/java/co/tpcreative/supersafe/common/services/SuperSafeApplication.java
package co.tpcreative.supersafe.common.services;
import android.Manifest;
import android.accounts.Account;
import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.support.multidex.MultiDex;
import android.support.multidex.MultiDexApplication;
import android.support.v4.content.PermissionChecker;
import android.util.Log;
import com.bumptech.glide.request.target.ViewTarget;
import com.crashlytics.android.Crashlytics;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.Scope;
import com.google.android.gms.drive.Drive;
import com.google.api.services.drive.DriveScopes;
import com.google.gson.Gson;
import com.snatik.storage.EncryptConfiguration;
import com.snatik.storage.Storage;
import com.snatik.storage.security.SecurityUtil;
import org.solovyev.android.checkout.Billing;
import co.tpcreative.supersafe.common.hiddencamera.config.CameraImageFormat;
import co.tpcreative.supersafe.common.util.Utils;
import co.tpcreative.supersafe.model.EnumPinAction;
import co.tpcreative.supersafe.model.EnumStatus;
import co.tpcreative.supersafe.model.room.InstanceGenerator;
import io.fabric.sdk.android.Fabric;
import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.annotation.Nonnull;
import co.tpcreative.supersafe.BuildConfig;
import co.tpcreative.supersafe.R;
import co.tpcreative.supersafe.common.api.RootAPI;
import co.tpcreative.supersafe.common.controller.ServiceManager;
import co.tpcreative.supersafe.common.controller.PrefsController;
import co.tpcreative.supersafe.common.network.Dependencies;
import co.tpcreative.supersafe.model.User;

public class SuperSafeApplication extends MultiDexApplication implements Dependencies.DependenciesListener, MultiDexApplication.ActivityLifecycleCallbacks {

    private static final String TAG = SuperSafeApplication.class.getSimpleName();
    private static SuperSafeApplication mInstance;
    private String supersafe;
    private String supersafePrivate;
    private String supersafeBackup;
    private String supersafeBreakInAlerts;
    private String supersafeLog;
    private String supersafeShare;
    private String supersafePicture;
    private String supersafeDataBaseFolder;
    private String key;
    private String fake_key;
    private String userSecret;
    private Storage storage;
    private EncryptConfiguration configurationFile;
    private EncryptConfiguration configurationPin;
    private static int resumed;
    private static int paused;
    private static int started;
    private static int stopped;
    private static String url;
    private int Orientation = 0;

    protected static Dependencies dependencies;
    public static RootAPI serverAPI;
    public static RootAPI serverDriveApi;
    public static RootAPI serviceGraphMicrosoft;
    private String authorization = null;

    private GoogleSignInOptions.Builder options;
    private Set<Scope> requiredScopes;
    private List<String> requiredScopesString;
    private boolean isLive = false;
    private String secretKey;
    private Activity activity;


    @Override
    public void onCreate() {
        super.onCreate();
        InstanceGenerator.getInstance(this);
        mInstance = this;
        isLive = false;

        Fabric.with(this, new Crashlytics());
        ViewTarget.setTagId(R.id.fab_glide_tag);

        /*Init own service api*/

        dependencies = Dependencies.getsInstance(getApplicationContext(), getUrl());
        dependencies.dependenciesListener(this);
        dependencies.init();
        serverAPI = (RootAPI) Dependencies.serverAPI;

        /*Init Drive api*/
        serverDriveApi = new RetrofitHelper().getCityService(getString(R.string.url_google));

        /*Init GraphMicrosoft*/
        serviceGraphMicrosoft = new RetrofitHelper().getCityService(getString(R.string.url_graph_microsoft));

        ServiceManager.getInstance().setContext(this);

        new PrefsController.Builder()
                .setContext(getApplicationContext())
                .setMode(ContextWrapper.MODE_PRIVATE)
                .setPrefsName(getPackageName())
                .setUseDefaultSharedPreference(true)
                .build();
        PrefsController.putInt(getString(R.string.key_screen_status), EnumPinAction.NONE.ordinal());
        PrefsController.putLong(getString(R.string.key_seek_to),0);
        PrefsController.putInt(getString(R.string.key_lastWindowIndex),0);

        /*Config file*/
        configurationFile = new EncryptConfiguration.Builder()
                .setChuckSize(1024*2)
                .setEncryptContent(SecurityUtil.IVX,getSecretKey(), SecurityUtil.SALT)
                .build();

        configurationPin = new EncryptConfiguration.Builder()
                .setChuckSize(1024*2)
                .setEncryptContent(SecurityUtil.IVX,SecurityUtil.SECRET_KEY, SecurityUtil.SALT)
                .build();

        storage = new Storage(getApplicationContext());
        supersafe = storage.getExternalStorageDirectory() + "/.SuperSafe_DoNot_Delete/";


        key = ".encrypt_key";
        fake_key = ".encrypt_fake_key";
        userSecret = ".userSecret";
        supersafePrivate = supersafe + "private/";
        supersafeBackup = supersafe + "backup/";
        supersafeLog = supersafe + "log/";
        supersafeBreakInAlerts = supersafe + "break_in_alerts/";
        supersafeShare = supersafe + "share/";
        supersafeDataBaseFolder = "/data/data/"+SuperSafeApplication.getInstance().getPackageName()+"/databases/";

        supersafePicture = storage.getExternalStorageDirectory(Environment.DIRECTORY_PICTURES) + "/SuperSafeExport/";
        registerActivityLifecycleCallbacks(this);
        Log.d(TAG, supersafe);


        options = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_client_id))
                .requestScopes(Drive.SCOPE_FILE)
                .requestScopes(Drive.SCOPE_APPFOLDER);

        requiredScopes = new HashSet<>(2);
        requiredScopes.add(Drive.SCOPE_FILE);
        requiredScopes.add(Drive.SCOPE_APPFOLDER);

        requiredScopesString = new ArrayList<>();
        requiredScopesString.add(DriveScopes.DRIVE_APPDATA);
        requiredScopesString.add(DriveScopes.DRIVE_FILE);

    }

    public String getSecretKey() {
        final User user = User.getInstance().getUserInfo();
        if (user!=null){
            if (user._id!=null){
                Utils.Log(TAG,"Get secret key " + user._id);
                secretKey = user._id;
                return secretKey;
            }
            Utils.Log(TAG,"secret id is null");
        }
        else{
            Utils.Log(TAG,"Get secret key null");
        }
        return SecurityUtil.SECRET_KEY;
    }

    public GoogleSignInOptions getGoogleSignInOptions(final Account account) {
        if (options != null) {
            if (account != null) {
                options.setAccount(account);
            }
            return options.build();
        }
        return options.build();
    }

    public EncryptConfiguration getConfigurationFile() {
        configurationFile = new EncryptConfiguration.Builder()
                .setChuckSize(1024*2)
                .setEncryptContent(SecurityUtil.IVX,getSecretKey(), SecurityUtil.SALT)
                .build();
        return configurationFile;
    }

    public List<String> getRequiredScopesString() {
        return requiredScopesString;
    }

    public Set<Scope> getRequiredScopes() {
        return requiredScopes;
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    public static synchronized SuperSafeApplication getInstance() {
        return mInstance;
    }

    /*In app purchase*/

    private final Billing mBilling = new Billing(this, new Billing.DefaultConfiguration() {
        /*In app purchase*/
        String key_purchase = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAk+6HXAFTNx3LbODafbpgsLqkdyMqMEvIYt55lqTjLIh0PkoAX7oSAD0fY7BXW0Czuys13hNNdyzmDjQe76xmUWTNfXM1vp0JQtStl7tRqNaFuaRje59HKRLpRTW1MGmgKw/19/18EalWTjbGOW7C2qZ5eGIOvGfQvvlraAso9lCTeEwze3bmGTc7B8MOfDqZHETdavSVgVjGJx/K10pzAauZFGvZ+ryZtU0u+9ZSyGx1CgHysmtfcZFKqZLbtOxUQHpBMeJf2M1LReqbR1kvJiAeLYqdOMWzmmNcsEoG6g/e+F9ZgjZjoQzqhWsrTE2IQZAaiwU4EezdqqruNXx6uwIDAQAB";
        @Override
        public String getPublicKey() {
            return key_purchase;
        }
    });

    @Nonnull
    public Billing getBilling() {
        return mBilling;
    }


    public void setConnectivityListener(SuperSafeReceiver.ConnectivityReceiverListener listener) {
        SuperSafeReceiver.connectivityReceiverListener = listener;
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle bundle) {
        if (this.activity==null){
            this.activity = activity;
        }
    }

    public Activity getActivity() {
        return activity;
    }

    @Override
    public void onActivityResumed(Activity activity) {
        ++resumed;
    }

    @Override
    public void onActivityPaused(Activity activity) {
        ++paused;
        Log.d(TAG, "application is in foreground: " + (resumed > paused));
    }

    @Override
    public void onActivityStarted(Activity activity) {
        ++started;
        Utils.Log(TAG, "onActivityStarted");
    }

    @Override
    public void onActivityStopped(Activity activity) {
        ++stopped;
        Log.d(TAG, "application is visible: " + (started > stopped));
    }

    public boolean isApplicationVisible() {
        return started > stopped;
    }

    public boolean isApplicationInForeground() {
        return resumed > paused;
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle bundle) {
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
    }

    public String getSuperSafe() {
        return supersafe;
    }

    public String getSupersafePrivate() {
        return supersafePrivate;
    }

    public String getSupersafeBackup() {
        return supersafeBackup;
    }

    public String getSupersafeLog() {
        return supersafeLog;
    }

    public String getSupersafeBreakInAlerts() {
        return supersafeBreakInAlerts;
    }

    public String getSupersafeShare() {
        return supersafeShare;
    }

    public String getSupersafePicture() {
        return supersafePicture;
    }

    public String getSupersafeDataBaseFolder() {
        return supersafeDataBaseFolder;
    }

    public void initFolder() {
        if (storage.isDirectoryExists(supersafe) & storage.isDirectoryExists(supersafePrivate) & storage.isDirectoryExists(supersafeBackup) & storage.isDirectoryExists(supersafeLog) & storage.isDirectoryExists(supersafeBreakInAlerts)) {
            Log.d(TAG, "SuperSafe is existing");
        } else {
            storage.createDirectory(supersafe);
            storage.createDirectory(supersafePrivate);
            storage.createDirectory(supersafeBackup);
            storage.createDirectory(supersafeLog);
            storage.createDirectory(supersafeBreakInAlerts);
            Log.d(TAG, "SuperSafe was created");
        }
    }

    public void deleteFolder() {
        storage.deleteDirectory(supersafe);
    }

    public void writeKey(String value) {
        if (!isPermissionWrite()) {
            Log.d(TAG, "Please grant access permission");
            return;
        }
        storage.setEncryptConfiguration(configurationPin);
        storage.createFile(getSuperSafe() + key, value);
        Log.d(TAG, "Created key :" + value);
    }

    public String readKey() {
        try {
            if (!isPermissionRead()) {
                Log.d(TAG, "Please grant access permission");
                return "";
            }
            storage.setEncryptConfiguration(configurationPin);
            boolean isFile = storage.isFileExist(getSuperSafe() + key);
            if (isFile) {
                String value = storage.readTextFile(getSuperSafe() + key);
                Log.d(TAG, "Key value is : " + value);
                return value;
            }
        }
        catch (Exception e){
            deleteFolder();
        }
        return "";

    }

    public void writeUserSecret(final User user) {
        if (!isPermissionWrite()) {
            Log.d(TAG, "Please grant access permission");
            return;
        }
        storage.setEncryptConfiguration(configurationPin);
        storage.createFile(getSuperSafe() + userSecret, new Gson().toJson(user));
    }

    public User readUseSecret() {
        try {
            if (!isPermissionRead()) {
                Log.d(TAG, "Please grant access permission");
                return null;
            }
            storage.setEncryptConfiguration(configurationPin);
            boolean isFile = storage.isFileExist(getSuperSafe() + userSecret);
            if (isFile) {
                String value = storage.readTextFile(getSuperSafe() + userSecret);
                if (value!=null){
                    Utils.Log(TAG,value);
                    final User mUser = new Gson().fromJson(value,User.class);
                    if (mUser!=null){
                        return mUser;
                    }
                }
                return null;
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    public void writeFakeKey(String value) {
        if (!isPermissionWrite()) {
            Log.d(TAG, "Please grant access permission");
            return;
        }
        storage.setEncryptConfiguration(configurationPin);
        storage.createFile(getSuperSafe() + fake_key, value);
        Log.d(TAG, "Created key :" + value);
    }

    public String readFakeKey() {
        if (!isPermissionRead()) {
            Log.d(TAG, "Please grant access permission");
            return "";
        }
        storage.setEncryptConfiguration(configurationPin);
        boolean isFile = storage.isFileExist(getSuperSafe() + fake_key);
        if (isFile) {
            String value = storage.readTextFile(getSuperSafe() + fake_key);
            Log.d(TAG, "Key value is : " + value);
            return value;
        }
        return "";
    }

    public void onDeleteKey() {
        if (!isPermissionRead()) {
            Log.d(TAG, "Please grant access permission");
            return;
        }
        boolean isFile = storage.isFileExist(getSuperSafe() + key);
        if (isFile) {
            storage.deleteFile(getSuperSafe() + key);
        }
    }

    public boolean isPermissionRead() {
        int permission = PermissionChecker.checkSelfPermission(getApplicationContext(), Manifest.permission.READ_EXTERNAL_STORAGE);
        if (permission == PermissionChecker.PERMISSION_GRANTED) {
            return true;
        }
        return false;
    }

    public boolean isPermissionWrite() {
        int permission = PermissionChecker.checkSelfPermission(getApplicationContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE);
        if (permission == PermissionChecker.PERMISSION_GRANTED) {
            return true;
        }
        return false;
    }

    public boolean isGrantAccess() {
        if (isPermissionWrite() && isPermissionRead()) {
            return true;
        }
        return false;
    }

    public int getOrientation() {
        return Orientation;
    }

    public void setOrientation(int mOrientation) {
        this.Orientation = mOrientation;
    }

    public Storage getStorage() {
        return storage;
    }

    public String getUrl() {
        if (!BuildConfig.DEBUG || isLive) {
            url = SecurityUtil.url_live;
        } else {
            url = SecurityUtil.url_developer;
        }
        return url;
    }

    public String getPathDatabase() {
        String currentDBPath = getDatabasePath(getString(R.string.key_database)).getAbsolutePath();
        return currentDBPath;
    }

    /*Retrofit and RXJava*/

    @Override
    public Class onObject() {
        return RootAPI.class;
    }

    @Override
    public String onAuthorToken() {
        try {
            String value = PrefsController.getString(getString(R.string.key_user), "");
            User user = new Gson().fromJson(value, User.class);
            if (user != null) {
                authorization = user.author.session_token;
                Utils.onWriteLog(authorization,EnumStatus.REQUEST_ACCESS_TOKEN);
            }
            else{
                user = readUseSecret();
                authorization = user.author.session_token;
            }
            return authorization;
        } catch (Exception e) {

        }
        return SecurityUtil.DEFAULT_TOKEN;
    }

    @Override
    public HashMap<String, String> onCustomHeader() {
        HashMap<String, String> hashMap = new HashMap<>();
        hashMap.put("Content-Type", "application/json");
        if (authorization != null) {
            hashMap.put("Authorization", authorization);
        }
        return hashMap;
    }

    @Override
    public boolean isXML() {
        return false;
    }

    public String getDeviceId() {
        return Settings.Secure.getString(getApplicationContext().getContentResolver(), Settings.Secure.ANDROID_ID);
    }

    public String getManufacturer() {
        String manufacturer = Build.MANUFACTURER;
        return manufacturer;
    }

    public String getModel() {
        String model = Build.MODEL;
        return model;
    }

    public int getVersion() {
        int version = Build.VERSION.SDK_INT;
        return version;
    }

    public String getVersionRelease() {
        String versionRelease = Build.VERSION.RELEASE;
        return versionRelease;
    }

    public  String getAppVersionRelease(){
        return BuildConfig.VERSION_NAME;
    }

    public File getPackagePath(Context context) {
        File file = new File(context.getExternalFilesDir(Environment.DIRECTORY_PICTURES),
                ".temporary.jpg");
        return file;
    }

    public File getPackageFolderPath(Context context) {
        File file = new File(context.getExternalFilesDir(Environment.DIRECTORY_PICTURES).getAbsolutePath());
        return file;
    }

    public File getDefaultStorageFile(int mImageFormat) {
        return new File(SuperSafeApplication.getInstance().getSupersafeBreakInAlerts()
                + File.separator
                + "IMG_" + System.currentTimeMillis()   //IMG_214515184113123.png
                + (mImageFormat == CameraImageFormat.FORMAT_JPEG ? ".jpeg" : ".png"));
    }

    public Map<String, String> getKeyHomePressed() {
        try {
            String value = PrefsController.getString(getString(R.string.key_home_pressed), null);
            if (value != null) {
                Map<String, String> map = new HashMap<String, String>();
                map = new Gson().fromJson(value, map.getClass());
                return map;
            }
            return new HashMap<>();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return new HashMap<>();
    }

    public void writeKeyHomePressed(String value) {
        try {
            Map<String, String> map = getKeyHomePressed();
            map.put(value, value);
            PrefsController.putString(getString(R.string.key_home_pressed), new Gson().toJson(map));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/co/tpcreative/supersafe/common/services/SuperSafeApplication.kt
package co.tpcreative.supersafe.common.services
import android.Manifest
import android.accounts.Account
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.ContextWrapper
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import androidx.core.content.PermissionChecker
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.multidex.MultiDex
import androidx.multidex.MultiDexApplication
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import co.tpcreative.supersafe.BuildConfig
import co.tpcreative.supersafe.R
import co.tpcreative.supersafe.common.api.ApiService
import co.tpcreative.supersafe.common.api.RetrofitBuilder
import co.tpcreative.supersafe.common.controller.PrefsController
import co.tpcreative.supersafe.common.encypt.SecurityUtil
import co.tpcreative.supersafe.common.extension.*
import co.tpcreative.supersafe.common.helper.EncryptDecryptFilesHelper
import co.tpcreative.supersafe.common.helper.EncryptDecryptPinHelper
import co.tpcreative.supersafe.common.helper.ThemeHelper
import co.tpcreative.supersafe.common.util.Utils
import co.tpcreative.supersafe.model.EnumPinAction
import co.tpcreative.supersafe.model.EnumThemeModel
import co.tpcreative.supersafe.model.EnumTypeServices
import co.tpcreative.supersafe.model.MigrationModel
import co.tpcreative.supersafe.ui.enterpin.EnterPinAct
import com.bumptech.glide.request.target.ImageViewTarget
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.Scope
import com.google.api.services.drive.DriveScopes
import com.google.firebase.crashlytics.FirebaseCrashlytics
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import java.io.*
import java.util.*

class SuperSafeApplication : MultiDexApplication(), Application.ActivityLifecycleCallbacks, LifecycleObserver {
    private lateinit var superSafe: String
    private lateinit var superSafeOldPath: String
    private lateinit var superSafePrivate: String
    private lateinit var superSafeBreakInAlerts: String
    private lateinit var superSafeLog: String
    private lateinit var superSafeShare: String
    private lateinit var superSafePicture: String
    private var key: String? = null
    private var fakeKey: String? = null
    private lateinit var options: GoogleSignInOptions.Builder
    private lateinit var requiredScopesString: MutableList<String>
    private var isLive = false
    private var activity: Activity? = null
    private var mMapMigrationItem: MutableMap<String, MigrationModel> = HashMap<String, MigrationModel>()
    private val classes = Stack<Activity>()

    override fun onCreate() {
        super.onCreate()
        initData()
        ProcessLifecycleOwner.get().lifecycle.addObserver(this);
    }

    fun initData(){
        mInstance = this
        isLive = true
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true)
        ImageViewTarget.setTagId(R.id.fab_glide_tag)
        serverApiCor = RetrofitBuilder.getService(typeService = EnumTypeServices.SYSTEM)
        serverDriveApiCor = RetrofitBuilder.getService(getString(R.string.url_google), typeService = EnumTypeServices.GOOGLE_DRIVE)
        serverMicCor = RetrofitBuilder.getService(getString(R.string.url_graph_microsoft), typeService = EnumTypeServices.EMAIL_OUTLOOK)
        PrefsController.Builder()
                .setContext(applicationContext)
                .setMode(ContextWrapper.MODE_PRIVATE)
                .setPrefsName(packageName)
                .setUseDefaultSharedPreference(true)
                .build()
        Utils.putScreenStatus(EnumPinAction.NONE.ordinal)
        Utils.putSeekTo(0)
        Utils.putLastWindowIndex(0)
        /*Migration*/
        if (isLiveMigration()) {
            superSafe = getExternalFilesDir(null)?.absolutePath + "/.SuperSafe_DoNot_Delete/"
            superSafeOldPath = "".getExternalStorageDirectory() + "/.SuperSafe_DoNot_Delete/"
        } else {
            superSafe = "".getExternalStorageDirectory() + "/.SuperSafe_DoNot_Delete/"
            superSafeOldPath = "".getExternalStorageDirectory() + "/.SuperSafe_DoNot_Delete/"
        }

        if (isLiveMigration()){
            key = superSafe + SecurityUtil.new_encrypt_key
            fakeKey = superSafe +SecurityUtil.new_encrypt_fake_Key
        }else{
            key = superSafe + SecurityUtil.old_encrypt_key
            fakeKey = superSafe +SecurityUtil.old_encrypt_fake_Key
        }

        superSafePrivate = superSafe + "private/"
        superSafeLog = superSafe + "log/"
        superSafeBreakInAlerts = superSafe + "break_in_alerts/"
        superSafeShare = superSafe + "share/"
        superSafePicture = "".getExternalStorageDirectory(Environment.DIRECTORY_DOWNLOADS) + "/SuperSafeExport/"
        registerActivityLifecycleCallbacks(this)
        Utils.Log(TAG, superSafe)
        options = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_client_id))
                .requestScopes(Scope(DriveScopes.DRIVE_FILE))
                .requestScopes(Scope(DriveScopes.DRIVE_APPDATA))
        requiredScopesString = ArrayList()
        requiredScopesString.add(DriveScopes.DRIVE_APPDATA)
        requiredScopesString.add(DriveScopes.DRIVE_FILE)
        Utils.onCheckNewVersion()
        ThemeHelper.applyTheme(EnumThemeModel.byPosition(Utils.getPositionThemeMode()))
        Utils.Log(TAG, getDeviceId())
    }

    fun getGoogleSignInOptions(account: Account?): GoogleSignInOptions? {
        if (account != null) {
            options.setAccountName(account.name)
        }
        return options.build()
    }

    fun getRequiredScopesString(): MutableList<String> {
        return requiredScopesString
    }

    override fun attachBaseContext(base: Context?) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    fun setConnectivityListener(listener: SuperSafeReceiver.ConnectivityReceiverListener?) {
        SuperSafeReceiver.connectivityReceiverListener = listener
    }

    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
        if (this.activity == null) {
            this.activity = activity
            Utils.Log(TAG, "onActivityCreated")
        }
        if (activity is EnterPinAct){
            classes.add(activity)
        }
    }

    override fun onActivityStarted(p0: Activity) {
        Utils.Log(TAG, "onActivityStarted $isRunningBackground")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    private fun onAppBackgrounded() {
        Utils.Log(TAG, "App in background")
        isRunningBackground = true
        Utils.putScreenStatus(EnumPinAction.SCREEN_LOCK.ordinal)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    private fun onAppForegrounded() {
        isRunningBackground  = false
        Utils.Log(TAG, "App in foreground")
    }

    override fun onActivityResumed(p0: Activity) {
        Utils.Log(TAG, "onActivityResumed")
    }

    override fun onActivityPaused(p0: Activity) {
        Utils.Log(TAG, "onActivityPaused")
    }

    override fun onActivityStopped(p0: Activity) {
        Utils.Log(TAG, "onActivityStopped")
    }

    fun getActivity(): Activity? {
        return activity
    }

    override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
    fun getSuperSafe(): String {
        return superSafe
    }

    fun getSuperSafeOldPath(): String {
        return superSafeOldPath
    }

    fun getSuperSafePrivate(): String {
        return superSafePrivate
    }

    fun getSuperSafeLog(): String {
        return superSafeLog
    }

    fun getFileLogs(): String {
        return "$superSafe/log.txt"
    }

    fun getSuperSafeBreakInAlerts(): String {
        return superSafeBreakInAlerts
    }

    fun getSuperSafeShare(): String {
        return superSafeShare
    }

    fun getSuperSafePicture(): String {
        return superSafePicture
    }

    fun initFolder() {
        if (superSafe.isDirectoryExists() and superSafePrivate.isDirectoryExists() and  superSafeLog.isDirectoryExists() and superSafeBreakInAlerts.isDirectoryExists() and superSafeOldPath.isDirectoryExists() and superSafePicture.isDirectoryExists()) {
            Utils.Log(TAG, "SuperSafe is existing")
        } else {
            if (isLiveMigration()){
                superSafe.createDirectory()
            }else{
                superSafeOldPath.createDirectory()
            }
            superSafePrivate.createDirectory()
            superSafeLog.createDirectory()
            superSafeBreakInAlerts.createDirectory()
            superSafePicture.createDirectory()
            Utils.Log(TAG, "SuperSafe was created")
        }
    }

    fun deleteFolder() {
        Utils.Log(TAG, "Request delete folder")
        superSafe.deleteDirectory()
    }

    fun writeKey(value: String?) {
        if (!isPermissionWrite()) {
            Utils.Log(TAG, "Please grant access permission")
            return
        }
        if (isLiveMigration()){
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_key
            if (oldPath.isFileExist()){
                if (EncryptDecryptFilesHelper.getInstance()?.createFile(key, value)==true){
                    oldPath.deleteFile()
                }
            }else{
                EncryptDecryptFilesHelper.getInstance()?.createFile(key, value)
            }
        }else{
            EncryptDecryptPinHelper.getInstance()?.createFile(key, value)
        }
        Utils.Log(TAG, "Created key :$value")
    }

    fun readKey(): String? {
        if (!isPermissionRead()) {
            Utils.Log(TAG, "Please grant access permission")
            return ""
        }
        if (isLiveMigration()){
            /*Checking old version*/
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_key
            if (oldPath.isFileExist()){
                return EncryptDecryptPinHelper.getInstance()?.readTextFile(oldPath)
            }
            return EncryptDecryptFilesHelper.getInstance()?.readTextFile(key)
        }else{
            return EncryptDecryptPinHelper.getInstance()?.readTextFile(key)
        }
    }

    fun writeFakeKey(value: String?) {
        if (!isPermissionWrite()) {
            Utils.Log(TAG, "Please grant access permission")
            return
        }
        if (isLiveMigration()){
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_fake_Key
            if (oldPath.isFileExist()){
                Utils.Log(TAG, "Deleted file")
                if (EncryptDecryptFilesHelper.getInstance()?.createFile(fakeKey, value) == true){
                    oldPath.deleteFile()
                }
            }else{
                EncryptDecryptFilesHelper.getInstance()?.createFile(fakeKey, value)
            }
        }else{
            EncryptDecryptPinHelper.getInstance()?.createFile(fakeKey, value)
        }
        Utils.Log(TAG, "Created key :$value")
    }

    fun readFakeKey() : String? {
        if (!isPermissionRead()) {
            Utils.Log(TAG, "Please grant access permission")
            return ""
        }
        if (isLiveMigration()){
            /*Checking old version*/
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_fake_Key
            if (oldPath.isFileExist()){
                return EncryptDecryptPinHelper.getInstance()?.readTextFile(oldPath)
            }
            return EncryptDecryptFilesHelper.getInstance()?.readTextFile(fakeKey)
        }else{
            return EncryptDecryptPinHelper.getInstance()?.readTextFile(fakeKey)
        }
    }

    private fun isPermissionRead(): Boolean {
        val permission: Int = PermissionChecker.checkSelfPermission(applicationContext, Manifest.permission.READ_EXTERNAL_STORAGE)
        return permission == PermissionChecker.PERMISSION_GRANTED
    }

    private fun isPermissionWrite(): Boolean {
        val permission: Int = PermissionChecker.checkSelfPermission(applicationContext, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        return permission == PermissionChecker.PERMISSION_GRANTED
    }

    fun isGrantAccess(): Boolean {
        return isPermissionWrite() && isPermissionRead()
    }

    fun getUrl(): String? {
        url = if (!BuildConfig.DEBUG || isLive) {
            SecurityUtil.url_live
        } else {
            SecurityUtil.url_developer
        }
        return url
    }

    fun getDeviceId(): String {
        val id: String? = Settings.Secure.getString(contentResolver, Settings.Secure.ANDROID_ID)
        return id ?: "${System.currentTimeMillis()}"
    }

    fun getManufacturer(): String? {
        return Build.MANUFACTURER
    }

    fun getModel(): String? {
        return Build.MODEL
    }

    fun getVersion(): Int {
        return Build.VERSION.SDK_INT
    }

    fun getVersionRelease(): String? {
        return Build.VERSION.RELEASE
    }

    fun getPackageId(): String {
        return BuildConfig.APPLICATION_ID
    }

    companion object {
        private val TAG = SuperSafeApplication::class.java.simpleName
        private var url: String? = null
        var serverApiCor : ApiService? = null
        var serverDriveApiCor : ApiService? = null
        var serverMicCor : ApiService? = null
        var isRunningBackground : Boolean? = false
        @Volatile
        private var mInstance: SuperSafeApplication? = null
        fun getInstance(): SuperSafeApplication {
            return mInstance as SuperSafeApplication
        }
    }

    val migrationFrom4To5: Migration = object : Migration(4, 5) {
        override fun migrate(database: SupportSQLiteDatabase) {
            database.execSQL("ALTER TABLE 'items' ADD COLUMN  'isUpdate' INTEGER NOT NULL DEFAULT 0")
        }
    }

    val migrationFrom5To6: Migration = object : Migration(5, 6) {
        override fun migrate(database: SupportSQLiteDatabase) {
            database.execSQL("ALTER TABLE 'items' ADD COLUMN  'isRequestChecking' INTEGER NOT NULL DEFAULT 0")
        }
    }

    val migrationFrom6To7: Migration = object : Migration(6, 7) {
        override fun migrate(database: SupportSQLiteDatabase) {
            database.execSQL("ALTER TABLE 'maincategories' ADD COLUMN  'created_date' TEXT DEFAULT '0'")
            database.execSQL("ALTER TABLE 'maincategories' ADD COLUMN  'updated_date' TEXT DEFAULT '0'")
            database.execSQL("ALTER TABLE 'maincategories' ADD COLUMN  'date_time' TEXT  DEFAULT '0'")
        }
    }

    suspend fun onPreparingMigration() = withContext(Dispatchers.IO) {
        mMapMigrationItem.clear()
        File(getSuperSafeOldPath()).walkTopDown().forEach {
            val mResult = it.absolutePath.replace(superSafeOldPath, superSafe)
            if (it.isFile) {
                val mUniqueId = UUID.randomUUID().toString()
                mMapMigrationItem[mUniqueId] = MigrationModel(mUniqueId, it, File(mResult), false)
                Utils.Log(TAG, "path of new file : $mResult")
            } else {
                mResult.createDirectory()
            }
        }
        Utils.Log(TAG,"Preparing migration....")
        with(mMapMigrationItem){
            forEach {
                val mValue = it.value
                moveTo(mValue.mInput, mValue.mOutput)
                responseMigration.invoke(mMapMigrationItem.size)
                Utils.Log(TAG, "loop of index ${mValue.Id}")
            }
        }
    }

    lateinit var responseMigration : (value: Int?) -> Unit

    private suspend fun moveTo(source: File, dest: File) = withContext(Dispatchers.IO){
        try {
            if (source.isDirectory) {
                source.mkdirs()
                Utils.Log(TAG, "Call here")
            }
            val fis = FileInputStream(source)
            val bufferLength = 1024
            val buffer = ByteArray(bufferLength)
            val fos = FileOutputStream(dest)
            val bos = BufferedOutputStream(fos, bufferLength)
            var read = 0
            read = fis.read(buffer, 0, read)
            while (read != -1) {
                bos.write(buffer, 0, read)
                read = fis.read(buffer) // if read value is -1, it escapes loop.
            }
            fis.close()
            bos.flush()
            bos.close()
            Utils.Log(TAG, "Finish...")
            if (!source.delete()) {
                Utils.Log(TAG, "failed to delete ${source.name}")
            }
        } catch (exception: IOException) {
            exception.printStackTrace()
            Utils.Log(TAG, "Could not move file...")
        }
    }

    fun isRequestMigration(): Boolean {
        File(getSuperSafeOldPath()).walkTopDown().forEach {
            if (it.isFile) {
                return true
            }
        }
        return false
    }

    fun isLiveMigration(): Boolean {
        if (!BuildConfig.DEBUG){
            return true
        }
        return true
    }

    fun checkingMigrationAfterVerifiedPin(value: String?, isReal: Boolean, isWrite: Boolean){
        if (isReal){
            if (checkingToWrite(isReal, isWrite)){
                writeKey(value)
            }
        }else{
            if (checkingToWrite(isReal, isWrite)){
                writeFakeKey(value)
            }
        }
    }

    private fun checkingToWrite(isReal: Boolean, isWrite: Boolean) : Boolean{
        if (isReal){
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_key
            if (oldPath.isFileExist()){
                return true
            }
            return isWrite
        }
        else{
            val oldPath = getSuperSafe() + SecurityUtil.old_encrypt_fake_Key
            if (oldPath.isFileExist()){
                return true
            }
            return isWrite
        }
    }

    fun isDebugPremium(): Boolean {
        if (BuildConfig.DEBUG){
            return true
        }
        return false
    }
}